### PR TITLE
Client / Server Auth Refactor

### PR DESCRIPTION
**Pull request by dbkr**
...including password reset.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/account.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            auth_user.to_string(),
+            threepid['medium'],
+            threepid['address'],
+            threepid['validated_at'],
+        )
+
+        if 'bind' in body and body['bind']:
+            logger.debug("Binding emails %s to %s" % (
+                threepid, auth_user.to_string()
+            ))
```
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/account.py]: You don't need to call "% ()" explicitly when logging:

    logging.debug("Binding emails %s to %s", threepid, auth_user.to_string())

Comment (dbkr) [inline synapse/rest/client/v2_alpha/account.py]: Ah, thanks for spotting.

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                logger.info("Auth completed with creds: %r", creds)
+                self._remove_session(sess)
+                defer.returnValue((True, creds, clientdict))
+
+        ret = self._auth_dict_for_flows(flows, sess)
+        ret['completed'] = creds.keys()
+        defer.returnValue((False, ret, clientdict))
+
+    @defer.inlineCallbacks
+    def add_oob_auth(self, stagetype, authdict, clientip):
```
Comment (NegativeMjark) [inline synapse/handlers/auth.py]: I think this could do with some doc string to explain what "add_oob_auth" does and what the return value means.

Comment (dbkr) [inline synapse/handlers/auth.py]: Done

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**